### PR TITLE
[vscode] parse view contribution "when" clause context

### PR DESCRIPTION
**Pull request by vinokurig**
#### What it does

There is an `affects()` function of the `ContextKeyChangeEvent`: https://github.com/theia-ide/theia/blob/b0d8292226e4ca179ed6cfa8be4e47e354ff9f61/packages/core/src/browser/context-key-service.ts#L34-L36
which evaluates if the given set of context keys has a context key that was changed.
`PluginViewRegistry` listens to this event and passes `when` clause context from the view contribution:
https://github.com/theia-ide/theia/blob/b0d8292226e4ca179ed6cfa8be4e47e354ff9f61/packages/plugin-ext/src/main/browser/view/plugin-view-registry.ts#L144
The`when` clause context from the view contribution is a string which may be a composed expression e.g. `"when": "!config.githubPullRequests.showInSCM && config.git.enabled"`: https://github.com/microsoft/vscode-pull-request-github/blob/d91e778705cd5af0785c5a05184ebc550334b409/package.json#L182

A map with parsed `when` clauses was introduced to pass separate context keys to the `affects()` function of the `ContextKeyChangeEvent`.

fixes https://github.com/theia-ide/theia/issues/6015
#### How to test

1. Remove `git` from `theia/examples/browser/package.json`.
2. Download [vscode git extension](https://github.com/che-incubator/vscode-git/releases/download/1.30.1/vscode-git-1.3.0.1.vsix).
3. Create `plugins` folder in Theia's root and paste the plugin to it.
4. Clone the [vscode github extension](https://github.com/microsoft/vscode-pull-request-github.git) and compile it.
5. Start the github plugin as a Hosted Plugin.
6. Click the GitHub icon.

Without this fix the GitHub panel is empty,
#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/browser/context-key-service.ts [hunk 1]
```diff
@@ -54,4 +59,11 @@ export class ContextKeyService {
         return true;
     }
 
+    /**
+     * It should be implemented by an extension, e.g. by the monaco extension.
+     */
+    deserialize(when: string): ContextKeyExpr {
```
Comment (akosyakov) [inline packages/core/src/browser/context-key-service.ts]: How about `Set<string>` as a return type?

**Comment by akosyakov** [comment]
@vinokurig It looks like a proper solution. 👍 I will test a bit later today.

File: packages/core/src/browser/context-key-service.ts (hunk 1 above)
Comment (akosyakov) [reply packages/core/src/browser/context-key-service.ts]: let's rename it to something else, like `parseKeys(when: string): Set<string>`

`ContextKeyExpr` does not look like complete api, and i don't think we want it to core right now

Comment (vinokurig) [reply packages/core/src/browser/context-key-service.ts]: done

**Comment by akosyakov** [comment]
@azatsarynnyy Could you check with vscode github extension please? I never was able to configure it properly. I will check that other extensions are still working.

**Comment by azatsarynnyy** [comment]
Sure. I'll check it.

File: packages/monaco/src/browser/monaco-context-key-service.ts [hunk 2]
```diff
@@ -58,4 +58,9 @@ export class MonacoContextKeyService extends ContextKeyService {
         return expression;
     }
 
+    parseKeys(expression: string): Set<string> {
+        const result = monaco.contextkey.ContextKeyExpr.deserialize(expression);
+        return new Set<string>(result.expr ? result.expr.map(e => e.key ? e.key : '') : []);
```
Comment (akosyakov) [inline packages/monaco/src/browser/monaco-context-key-service.ts]: `expr` can be recursive, so we should traverse it recursively and collect all keys

File: packages/monaco/src/typings/monaco/index.d.ts [hunk 3]
```diff
@@ -1119,6 +1119,8 @@ declare module monaco.contextKeyService {
 
 declare module monaco.contextkey {
     export class ContextKeyExpr {
+        expr?: ContextKeyExpr[];
+        key?: string;
         static deserialize(when: string): ContextKeyExpr;
```
Comment (akosyakov) [inline packages/monaco/src/typings/monaco/index.d.ts]: expression has already `keys` property: https://github.com/TypeFox/vscode/blob/70b8db24a37fafc77247de7f7cb5bb0195120ed0/src/vs/platform/contextkey/common/contextkey.ts#L161

We should expose and use it instead.

File: packages/monaco/src/browser/monaco-context-key-service.ts (hunk 2 above)
Comment (vinokurig) [reply packages/monaco/src/browser/monaco-context-key-service.ts]: fixed

File: packages/monaco/src/typings/monaco/index.d.ts (hunk 3 above)
Comment (vinokurig) [reply packages/monaco/src/typings/monaco/index.d.ts]: fixed

**Comment by vinokurig** [comment]
@akosyakov 
> It does now work for json outline sample extension: https://github.com/microsoft/vscode-extension-samples/tree/master/tree-view-sample

Looks like this controls the `json outline` in vscode navigator, but is there such view in theia?

**Comment by akosyakov** [comment]
@vinokurig it should be contributed by the sample extension, but it only appears when json file is opened and should disappear when it is closed

**Comment by azatsarynnyy** [comment]
I've checked it with GitHub Pull Requests extension.

`config.git.enabled` is `true` && `config.githubPullRequests.showInSCM` is `true`

image

`config.git.enabled` is `true` && `config.githubPullRequests.showInSCM` is `false`

image

`config.git.enabled` is `false` && `config.githubPullRequests.showInSCM` is `false`

image

**Comment by azatsarynnyy** [comment]
I'm not sure whether it's related to this PR but there's `Sign in` button on the view in VS Code.

image

I don't see it in Theia.

**Comment by azatsarynnyy** [comment]
It fixes the issue with GH PR extension.
`Sign in` button visibility is not in the scope of this issue.

**Pull request merged**